Source/ui_unix/GamePad/GamePadDeviceListener_OSX.cpp

```diff
@@ -0,0 +1,350 @@
+	}
+}
+void CGamePadDeviceListener::SetInitialBindValues(CGamePadDeviceListener* GPDL, IOHIDDeviceRef device)
+{
+	CFArrayRef elements = IOHIDDeviceCopyMatchingElements(device, nullptr, 0);
+
+	for(int i = 0; i < CFArrayGetCount(elements); i++)
+	{
+		IOHIDElementRef elementRef = (IOHIDElementRef)CFArrayGetValueAtIndex(elements, i);
+		uint32 usagePage = IOHIDElementGetUsagePage(elementRef);
+		if(usagePage == kHIDPage_VendorDefinedStart)
+		{
+			continue;
+		}
+		IOHIDValueRef valueRef;
+		if(IOHIDDeviceGetValue(device, elementRef, &valueRef) != kIOReturnSuccess)
+		{
+			continue;
+		}
+
+		CFIndex value = IOHIDValueGetIntegerValue(valueRef);
+		IOHIDElementType type = IOHIDElementGetType(elementRef);
+		uint32 usage = IOHIDElementGetUsage(elementRef);
+		int btn_type = btn_type::digital;
+		bool is_axis = type == kIOHIDElementTypeInput_Axis || type == kIOHIDElementTypeInput_Misc;
+		if(is_axis)
+		{
+			btn_type = btn_type::axis;
+			if(usage == kHIDUsage_GD_Hatswitch)
+			{
+				btn_type = btn_type::hatswitch;
+			}
+		}
+		switch(type)
+		{
+		case kIOHIDElementTypeInput_Misc:
+		case kIOHIDElementTypeInput_Button:
+		case kIOHIDElementTypeInput_Axis:
+			GPDL->OnInputEventCallBack(CGamePadUtils::GetDeviceID(device), usage, value, btn_type);
+			break;
+		default:
+			break;
+		}
+	}
+}
+IOHIDReportCallback CGamePadDeviceListener::GetCallback(CGamePadDeviceListener* GPDL, IOHIDDeviceRef device)
+{
+	auto vid = CGamePadUtils::GetIntProperty(device, CFSTR(kIOHIDVendorIDKey));
+	auto pid = CGamePadUtils::GetIntProperty(device, CFSTR(kIOHIDProductIDKey));
+	if(vid == 0x54C)
+	{
+		if(pid == 0x9CC || pid == 0x5C4)
+		{
+			return GPDL->InputReportCallbackStub_DS4;
+		}
+		else if(pid == 0x268)
+		{
+			return GPDL->InputReportCallbackStub_DS3;
+		}
+	}
+	return nullptr;
+}
+
+void CGamePadDeviceListener::InputDeviceListenerThread()
+{
+	m_hidManager = IOHIDManagerCreate(kCFAllocatorDefault, 0);
+	{
+		CFDictionaryRef matchingDict[3];
+		matchingDict[0] = CreateDeviceMatchingDictionary(kHIDPage_GenericDesktop, kHIDUsage_GD_Joystick);
+		matchingDict[1] = CreateDeviceMatchingDictionary(kHIDPage_GenericDesktop, kHIDUsage_GD_GamePad);
+		matchingDict[2] = CreateDeviceMatchingDictionary(kHIDPage_GenericDesktop, kHIDUsage_GD_MultiAxisController);
+
+		CFArrayRef array = CFArrayCreate(kCFAllocatorDefault, (const void**)matchingDict, 3, &kCFTypeArrayCallBacks);
+		CFRelease(matchingDict[0]);
+		CFRelease(matchingDict[1]);
+		CFRelease(matchingDict[2]);
+		IOHIDManagerSetDeviceMatchingMultiple(m_hidManager, array);
+	}
+
+	IOHIDManagerRegisterDeviceMatchingCallback(m_hidManager, onDeviceMatched, this);
+	//IOHIDManagerRegisterDeviceRemovalCallback(m_hidManager, onDeviceRemoved, this);
+
+	IOHIDManagerOpen(m_hidManager, kIOHIDOptionsTypeNone);
+	IOHIDManagerScheduleWithRunLoop(m_hidManager, CFRunLoopGetCurrent(), CFSTR("CustomLoop"));
+	while(CFRunLoopRunInMode(CFSTR("CustomLoop"), 0, true) != kCFRunLoopRunStopped && m_running)
```

Comment:
Specifying 0 for the seconds parameter in CFRunLoopRunInMode will cause 100% CPU usage. Using 1 instead of 0 seems to fix the problem.